### PR TITLE
ci: add Android job

### PR DESCRIPTION
**Pull request by mxinden**
Runs unit tests on Android emulator for both `x86` and `x86_64`.

---

Tracking issue: https://github.com/quinn-rs/quinn/issues/1778

Foundation to test https://github.com/quinn-rs/quinn/issues/1947.

### CODE REVIEW

File: quinn/src/tests.rs [hunk 1]
```diff
@@ -523,6 +523,7 @@ fn run_echo(args: EchoArgs) {
             // requires modifying this test - please update the list of supported
             // platforms in the doc comment of `quinn_udp::RecvMeta::dst_ip`.
```
Comment (mxinden) [inline quinn/src/tests.rs:524]: Note that Android has already been added here:

https://github.com/quinn-rs/quinn/blob/169f0c2b4380fa7fc4923d9a7a3058f4bc8939d2/quinn-udp/src/lib.rs#L91-L93

**Review by mxinden — COMMENTED** [review]
This is ready for a first review.

I hoped to be able to reproduce https://github.com/quinn-rs/quinn/issues/1947 with the `x86` branch of this workflow. Thus far without success. I will debug further. In the meantime, I believe running Android tests is useful by itself.

Let me know what you think.

File: .github/workflows/rust.yml [hunk 2]
```diff
@@ -91,3 +91,74 @@ jobs:
...
+
+    - name: Enable KVM group perms
+      run: |
+        echo 'KERNEL=="kvm", GROUP="kvm", MODE="0666", OPTIONS+="static_node=kvm"' | sudo tee /etc/udev/rules.d/99-kvm4all.rules
+        sudo udevadm control --reload-rules
+        sudo udevadm trigger --name-match=kvm
+
+    - name: Create emulator script
+      run: |
+        echo '#!/bin/bash
```
Comment (djc) [inline .github/workflows/rust.yml:144]: Don't love having this longish shell script inside the YAML. Maybe extract it into a separate file?

**Review by djc — APPROVED** [review]
This mostly looks good to me.

**Review by Ralith — APPROVED** [review]

File: .github/workflows/rust.yml [hunk 3]
```diff
@@ -91,3 +91,61 @@ jobs:
...
+      matrix:
+        include:
+          - target: x86_64-linux-android
+            emulator-arch: x86_64
+            # Note that x86_64 image is only available for API 21+. See
+            # https://github.com/ReactiveCircus/android-emulator-runner?tab=readme-ov-file#configurations.
+            api-level: 21
+          - target: i686-linux-android
+            emulator-arch: x86
+            api-level: 19
```
Comment (mxinden) [inline .github/workflows/rust.yml:108]: Uses Android API level 19 as suggested in https://github.com/quinn-rs/quinn/issues/1503#issuecomment-2285989144 by @link2xt.

File: .github/workflows/rust.yml (hunk 2 above)
Comment (mxinden) [reply .github/workflows/rust.yml]: Extracted into a separate file. Mind taking another look?

https://github.com/quinn-rs/quinn/pull/1950/commits/3fa87f2a2ae803b6b4d13a3314cd1997395dbd16

**Comment by mxinden** [comment]
```
error: Failed loading manifest
error: Could not derive library name from Cargo.toml
```

Haven't seen this happening before. I will look into it.

**Pull request merged**